src/NLog/Targets/FileTarget.cs

```diff
@@ -1966,12 +1983,50 @@ private string GetArchiveFileNameBasedOnTime(string fileName, LogEventInfo logEv
                 string formatString = GetArchiveDateFormatString(string.Empty);
                 string fileCreated = creationTimeSource.Value.ToString(formatString, CultureInfo.InvariantCulture);
                 string logEventRecorded = logEvent.TimeStamp.ToString(formatString, CultureInfo.InvariantCulture);
-
                 var shouldArchive = fileCreated != logEventRecorded;
                 if (shouldArchive)
                 {
-                    return fileName;
+                    FileInfo fileInfo = new FileInfo(fileName);
+                    if (fileInfo.Exists)
+                    {
+                        return fileName;
+                    }
+                    else
+                    {
+                        _fileAppenderCache.InvalidateAppender(fileName)?.Dispose(); // File deleted behind our backs
+                        _initializedFiles.Remove(fileName);
+                    }
+                }
+            }
+
+            return null;
+        }
+
+        private DateTime? TryGetArchiveFileCreationTimeSource(string fileName, DateTime previousLogEventTimestamp)
+        {
+            // Linux FileSystems doesn't always have file-birth-time, so NLog tries to provide a little help
+            DateTime? fallbackTimeSourceLinux = (previousLogEventTimestamp != DateTime.MinValue && IsSimpleKeepFileOpen) ? previousLogEventTimestamp : (DateTime?)null;
+            var creationTimeSource = _fileAppenderCache.GetFileCreationTimeSource(fileName, fallbackTimeSourceLinux);
+            var fileLength = creationTimeSource.HasValue ? _fileAppenderCache.GetFileLength(fileName) : null;   // Verifies file-handle by checking FileStream.Length
+            if (creationTimeSource.HasValue && fileLength.HasValue)
+            {
+                if (previousLogEventTimestamp != DateTime.MinValue && previousLogEventTimestamp < creationTimeSource)
+                {
+                    if (TruncateArchiveTime(previousLogEventTimestamp, FileArchivePeriod.Minute) < TruncateArchiveTime(creationTimeSource.Value, FileArchivePeriod.Minute) && PlatformDetector.IsUnix)
+                    {
+                        if (IsSimpleKeepFileOpen)
```

Comment:
Could we please improve the nesting here? For every uncombined `if` will sonarqube complain;)

I think we need some helper methods/variables

Reply:
Removed some of the nesting